### ISSUE TITLE
Motivation behind completely static build

### ISSUE THREAD

**Issue by gfiumara** [usnistgov/NFIQ2#21]
I can understand the desire for a completely static build, but also appreciate the option for a dynamically linked build. That said, I am not sure I understand the motivation behind statically linking the entire C and C++ libraries, [as performed here](https://github.com/usnistgov/NFIQ2/blob/ba0b03493bfcbc0b35c69aff0bfe5a704e9befc0/NFIQ2/compiler.cmake#L21).

**Comment by RLessmann** [comment]
Quite easy, NFIQ2 is approved with OpenCV2 and not with any newer version. Open CV is widely used within the industry and many application have this OpenCV support (unknown version) build in. In order to force the use of OpenCV 2 in NFIQ2 you better link it staticly to have no naming collisions within you application.
Also for a deployment with minimal dependencies it is preferable to link important libraries statically.

But the approach might differ here from user to user.

**Comment by gfiumara** [comment]
No issue with providing a version of NFIQ2 statically-linked against OpenCV. I agree this can make a lot of sense. I'd like the option for both static and dynamic link of OpenCV, but that is a separate discussion not worth holding up integration.

Statically linking against libstdc++ is different in my mind. I may be wrong, but I'm anticipating high likely hood of symbol mixups/undefined behavior when something dynamically linked needs to use a libstdc++ symbol if it was not built at the exact same time. The man page for g++ mentions that exception handling in shared libraries will have issues with a static libgcc.

**Comment by RLessmann** [comment]
This definitely needs further thoughts. If you link opencv statically you are already bound to a certain   libstdc++, if dynamicaly a wrong one is used you will see pretty nice crashes. 
IMHO static linkage is not preferred but definitely solves portability issues.

**Comment by gfiumara** [comment]
I see undefined references to symbols from glibc in OpenCV libraries and ultimately the nfiq2 binary, so maybe it's not as static as we think right now anyway. 

Right now, we're requiring people build the entire codebase themselves. Portability issues come into play if we distribute pre-built "official releases." In my opinion, that's something we should do, but for supported OS. Let's continue to discuss, but I want to be sure we're not going to introduce strange behavior with these compiler flags. I haven't used them personally, so some experimentation and reading are required on my end.

**Comment by RLessmann** [comment]
Probably a misunderstanding. IMHO glibc is the standard C, but not the standard C++. Standard C libs are fine dynamicaly to be used.
But agreed need more confidence on this one.

**Comment by gfiumara** [comment]
As discussed in person, this seemed to largely come about because of a desire to have reliable self-contained binaries that can be shipped and work on an OS of a particular type. This may be resolved by #42. Below is a list of libraries that are still linked dynamically. Most are image libraries, which are known for being the source of security vulnerabilities, and so it would make sense to continue to link all these dynamically.

```
$ ldd ./dist/Linux-x86_64/x64/nfiq2.GNU.linux64.ocv347.fjfx521.test 
linux-vdso.so.1 =>  
libdl.so.2 => /lib64/libdl.so.2 
libjpeg.so.62 => /lib64/libjpeg.so.62 
libpng15.so.15 => /lib64/libpng15.so.15 
libtiff.so.5 => /lib64/libtiff.so.5 
libpthread.so.0 => /lib64/libpthread.so.0 
librt.so.1 => /lib64/librt.so.1 
libm.so.6 => /lib64/libm.so.6 
libc.so.6 => /lib64/libc.so.6 
/lib64/ld-linux-x86-64.so.2 
libz.so.1 => /lib64/libz.so.1 
libjbig.so.2.0 => /lib64/libjbig.so.2.0 
```

**Comment by gfiumara** [comment]
Closing this issue. I think we understand the rationale behind builds that include the standard libraries statically.

We also plan to release pre-compiled binaries for different systems (#42) and switch between dynamic and static builds (#56).
